cpp/src/io/parquet/reader_impl.cpp

```diff
@@ -628,22 +616,25 @@ table_with_metadata reader::impl::read_chunk()
     }
   }
 
-  prepare_data(0 /*skip_rows*/,
-               std::nullopt /*num_rows, `nullopt` means unlimited*/,
-               true /*uses_custom_row_bounds*/,
-               {} /*row_group_indices, empty means read all row groups*/,
-               std::nullopt /*filter*/);
+  // Save the output filter for use in `finalize_output()`
+  table_metadata metadata;
+  populate_metadata(metadata);
+  auto expr_conv = named_to_reference_converter(_options.filter, metadata);
+  _output_filter = expr_conv.get_converted_expr();
 
-  return read_chunk_internal(true, std::nullopt);
+  prepare_data(read_mode::CHUNKED_READ);
+  return read_chunk_internal(read_mode::CHUNKED_READ);
 }
 
 bool reader::impl::has_next()
 {
-  prepare_data(0 /*skip_rows*/,
-               std::nullopt /*num_rows, `nullopt` means unlimited*/,
-               true /*uses_custom_row_bounds*/,
-               {} /*row_group_indices, empty means read all row groups*/,
-               std::nullopt /*filter*/);
+  // Save the output filter for use in `finalize_output()`
+  table_metadata metadata;
+  populate_metadata(metadata);
+  auto expr_conv = named_to_reference_converter(_options.filter, metadata);
+  _output_filter = expr_conv.get_converted_expr();
```

Comment:
Wait, this seems redundant/expensive as here is inside `has_next` which should not handle data output.

Reply:
It is, I mean to enclose it in an `if` clause after copy-pasting. Pushed the fix just now!

Reply:
Moved this piece of code in the constructor to simplify!